File: Modify SRT/Shift.rb
def shift_subtitle_time(subtitle, shift)
    start_time, end_time = subtitle[1], subtitle[2]
    shifted_start_time = shift_time(start_time, shift)
    shifted_end_time = shift_time(end_time, shift)
    "#{shifted_start_time} --> #{shifted_end_time}"
end

def shift_time(time_str, shift)
    hours, minutes, seconds, milliseconds = time_str.split(/[:,]/)
    total_milliseconds = hours.to_i*3600000 + minutes.to_i*60000 + seconds.to_i*1000 + milliseconds.to_i
    shifted_milliseconds = total_milliseconds + shift
    shifted_hours, remainder = shifted_milliseconds.divmod(3600000)
    shifted_minutes, remainder = remainder.divmod(60000)
    shifted_seconds, shifted_milliseconds = remainder.divmod(1000)
    "#{pad_zero(shifted_hours)}:#{pad_zero(shifted_minutes)}:#{pad_zero(shifted_seconds)},#{pad_zero(shifted_milliseconds, 3)}"
end

def pad_zero(value, length = 2)
    value.to_s.rjust(length, '0')
end

def main
    # Example usage
    input_file_path = "S01E01.srt"
    output_file_path = "S01E01_shifted.srt"
    time_shift = 2000  # Shift the subtitles 2 seconds forward

    File.open(input_file_path, 'r') do |input_file|
        File.open(output_file_path, 'w') do |output_file|
            input_file.each_line do |line|
                match = line.match(/(\d+:\d+:\d+,\d+) --> (\d+:\d+:\d+,\d+)/)
                if match
                    shifted_time_line = shift_subtitle_time(match, time_shift)
                    output_file.puts shifted_time_line
                else
                output_file.puts line
                end
            end
        end
    end
end

def clear_console
    if RUBY_PLATFORM =~ /linux|darwin/  # for Unix-like systems (Linux, macOS)
      system "clear"
    else  # for Windows
      system "cls"
    end
end

#start_time = Time.now #Used to record the execution time
main
#end_time = Time.now
#execution_time = end_time - start_time
#puts "Execution time: #{execution_time}"

File: Modify SRT/Shift.pl
#!/usr/bin/env perl
use strict;
#Use warnings;
#Use Term::ANSIColor;
use Time::HiRes qw(gettimeofday); #Used to record the execution time

sub shift_subtitle_time 
{
    my ($subtitle, $shift) = @_;
    my ($start_time, $end_time) = ($subtitle =~ /(\d+:\d+:\d+,\d+) --> (\d+:\d+:\d+,\d+)/);
    my $shifted_start_time = shift_time($start_time, $shift);
    my $shifted_end_time = shift_time($end_time, $shift);
    return "$shifted_start_time --> $shifted_end_time";
}

sub shift_time 
{
    my ($time_str, $shift) = @_;
    my ($hours, $minutes, $seconds, $milliseconds) = ($time_str =~ /(\d+):(\d+):(\d+),(\d+)/);
    my $total_milliseconds = $hours*3600000 + $minutes*60000 + $seconds*1000 + $milliseconds;
    my $shifted_milliseconds = $total_milliseconds + $shift;
    my ($shifted_hours, $remainder) = divmod($shifted_milliseconds, 3600000);
    my ($shifted_minutes, $remainder) = divmod($remainder, 60000);
    my ($shifted_seconds, $shifted_milliseconds) = divmod($remainder, 1000);
    return sprintf("%02d:%02d:%02d,%03d", $shifted_hours, $shifted_minutes, $shifted_seconds, $shifted_milliseconds);
}

sub divmod 
{
    my ($dividend, $divisor) = @_;
    my $quotient = int($dividend / $divisor);
    my $remainder = $dividend % $divisor;
    return ($quotient, $remainder);
}

sub main 
{
    #clear_console();
    # Example usage
    my $input_file_path = "S01E01.srt";
    my $output_file_path = "S01E01_shifted.srt";
    my $time_shift = 2000;  # Shift the subtitles 2 seconds forward

    open(my $input_file, '<', $input_file_path) or die "Cannot open input file: $!";
    open(my $output_file, '>', $output_file_path) or die "Cannot open output file: $!";

    while (my $line = <$input_file>) 
    {
        if (my ($start_time, $end_time) = ($line =~ /(\d+:\d+:\d+,\d+) --> (\d+:\d+:\d+,\d+)/)) {
            my $shifted_time_line = shift_subtitle_time($line, $time_shift);
            print $output_file $shifted_time_line . "\n";
        } else {
            print $output_file $line;
        }
    }
    close($input_file);
    close($output_file);
}

sub clear_console 
{
    if ($^O =~ /linux|darwin/) {  # for Unix-like systems (Linux, macOS)
        system "clear";
    } else {  # for Windows
        system "cls";
    }
}

#my $start_time = gettimeofday();
main();
#my $end_time = gettimeofday();
#my $execution_time = $end_time - $start_time;
#print "Execution time: $execution_time \n";